### PR TITLE
[LV][VPlan] Use VF VPValue in VPVectorPointerRecipe

### PR DESCRIPTION
**Pull request by arcbbb**
Refactors VPVectorPointerRecipe to use the VF VPValue to obtain the runtime VF, similar to #95305.

Since only reverse vector pointers require the runtime VF, the patch sets VPUnrollPart::PartOpIndex to 1 for vector pointers and 2 for reverse vector pointers. As a result, the generation of reverse vector pointers is moved into a separate recipe.

### CODE REVIEW

File: llvm/test/Transforms/LoopVectorize/RISCV/riscv-vector-reverse.ll [hunk 1]
```diff
@@ -53,41 +53,42 @@ define void @vector_reverse_i64(ptr nocapture noundef writeonly %A, ptr nocaptur
...
 ; CHECK-NEXT:  <x1> vector loop: {
 ; CHECK-NEXT:    vector.body:
-; CHECK-NEXT:      EMIT vp<%3> = CANONICAL-INDUCTION ir<0>, vp<%8>
-; CHECK-NEXT:      vp<%4> = DERIVED-IV ir<%n> + vp<%3> * ir<-1>
-; CHECK-NEXT:      vp<%5> = SCALAR-STEPS vp<%4>, ir<-1>
-; CHECK-NEXT:      CLONE ir<%i.0> = add nsw vp<%5>, ir<-1>
+; CHECK-NEXT:      EMIT vp<%4> = CANONICAL-INDUCTION ir<0>, vp<%9>
+; CHECK-NEXT:      vp<%5> = DERIVED-IV ir<%n> + vp<%4> * ir<-1>
+; CHECK-NEXT:      vp<%6> = SCALAR-STEPS vp<%5>, ir<-1>
+; CHECK-NEXT:      CLONE ir<%i.0> = add nsw vp<%6>, ir<-1>
```
Comment (fhahn) [inline llvm/test/Transforms/LoopVectorize/RISCV/riscv-vector-reverse.ll:74]: Would it be possible to update the test to use patterns for the VPValue numbers separately, to avoid the unrelated changes?

File: llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp [hunk 2]
```diff
@@ -1813,38 +1813,61 @@ void VPWidenGEPRecipe::print(raw_ostream &O, const Twine &Indent,
...
+  const DataLayout &DL = Builder.GetInsertBlock()->getDataLayout();
+  Type *IndexTy = State.VF.isScalable()
+                      ? DL.getIndexType(IndexedTy->getPointerTo())
+                      : Builder.getInt32Ty();
+  Value *Ptr = State.get(getOperand(0), VPLane(0));
+  bool InBounds = isInBounds();
+
+  // the wide store needs to start at the last vector element.
+  // RunTimeVF =  VScale * VF.getKnownMinValue()
+  // For fixed-width VScale is 1, then RunTimeVF = VF.getKnownMinValue()
```
Comment (fhahn) [inline llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:1831]: This seems out of place now?

File: llvm/test/Transforms/LoopVectorize/RISCV/riscv-vector-reverse.ll (hunk 1 above)
Comment (arcbbb) [reply llvm/test/Transforms/LoopVectorize/RISCV/riscv-vector-reverse.ll:74]: sure, separate it into https://github.com/llvm/llvm-project/pull/111086

File: llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp (hunk 2 above)
Comment (arcbbb) [reply llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp]: Removed. Thanks!

**Comment by arcbbb** [comment]
rebased to avoid unrelated changes on tests

File: llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp [hunk 3]
```diff
@@ -1813,38 +1813,59 @@ void VPWidenGEPRecipe::print(raw_ostream &O, const Twine &Indent,
...
+  // Use i32 for the gep index type when the value is constant,
+  // or query DataLayout for a more suitable index type otherwise.
+  const DataLayout &DL = Builder.GetInsertBlock()->getDataLayout();
+  Type *IndexTy = State.VF.isScalable()
+                      ? DL.getIndexType(IndexedTy->getPointerTo())
+                      : Builder.getInt32Ty();
+  Value *Ptr = State.get(getOperand(0), VPLane(0));
+  bool InBounds = isInBounds();
+
+  // the wide store needs to start at the last vector element.
```
Comment (fhahn) [inline llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:1829]:
```suggestion
  // The wide store needs to start at the last vector element.
```

File: llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp [hunk 4]
```diff
@@ -1813,38 +1813,59 @@ void VPWidenGEPRecipe::print(raw_ostream &O, const Twine &Indent,
...
+void VPReverseVectorPointerRecipe ::execute(VPTransformState &State) {
+  auto &Builder = State.Builder;
+  State.setDebugLocFrom(getDebugLoc());
+  unsigned CurrentPart = getUnrollPart(*this);
+  // Use i32 for the gep index type when the value is constant,
+  // or query DataLayout for a more suitable index type otherwise.
+  const DataLayout &DL = Builder.GetInsertBlock()->getDataLayout();
+  Type *IndexTy = State.VF.isScalable()
+                      ? DL.getIndexType(IndexedTy->getPointerTo())
+                      : Builder.getInt32Ty();
```
Comment (fhahn) [inline llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:1825]: Might be good to share this code between both vector pointer recipes?

File: llvm/lib/Transforms/Vectorize/VPlan.h [hunk 5]
```diff
@@ -1785,20 +1787,65 @@ class VPWidenGEPRecipe : public VPRecipeWithIRFlags {
...
+  /// Print the recipe.
+  void print(raw_ostream &O, const Twine &Indent,
+             VPSlotTracker &SlotTracker) const override;
+#endif
+};
+
 /// A recipe to compute the pointers for widened memory accesses of IndexTy for
-/// all parts. If IsReverse is true, compute pointers for accessing the input in
-/// reverse order per part.
+/// all parts.
```
Comment (fhahn) [inline llvm/lib/Transforms/Vectorize/VPlan.h:1838]: While you are here, drop the `for all parts` bit, as it is stale.

File: llvm/lib/Transforms/Vectorize/VPlan.h [hunk 6]
```diff
@@ -1785,20 +1787,65 @@ class VPWidenGEPRecipe : public VPRecipeWithIRFlags {
 #endif
 };
 
+/// A recipe to compute the pointers for widened memory accesses of IndexTy
+/// in reverse order per part.
```
Comment (fhahn) [inline llvm/lib/Transforms/Vectorize/VPlan.h:1791]:
```suggestion
/// in reverse order.
```

File: llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp [hunk 7]
```diff
@@ -1813,38 +1813,59 @@ void VPWidenGEPRecipe::print(raw_ostream &O, const Twine &Indent,
...
+  Value *Ptr = State.get(getOperand(0), VPLane(0));
+  bool InBounds = isInBounds();
```
Comment (fhahn) [inline llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:1827]: Sink to use

File: llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp [hunk 8]
```diff
@@ -1813,38 +1813,59 @@ void VPWidenGEPRecipe::print(raw_ostream &O, const Twine &Indent,
...
+  auto &Builder = State.Builder;
+  State.setDebugLocFrom(getDebugLoc());
+  unsigned CurrentPart = getUnrollPart(*this);
+  // Use i32 for the gep index type when the value is constant,
+  // or query DataLayout for a more suitable index type otherwise.
+  const DataLayout &DL = Builder.GetInsertBlock()->getDataLayout();
+  Type *IndexTy = State.VF.isScalable()
+                      ? DL.getIndexType(IndexedTy->getPointerTo())
+                      : Builder.getInt32Ty();
+  Value *Ptr = State.get(getOperand(0), VPLane(0));
```
Comment (fhahn) [inline llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:1826]: Sink to use

File: llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp [hunk 9]
```diff
@@ -1813,38 +1813,59 @@ void VPWidenGEPRecipe::print(raw_ostream &O, const Twine &Indent,
...
+
+  State.set(this, ResultPtr, /*IsScalar*/ true);
+}
+
+#if !defined(NDEBUG) || defined(LLVM_ENABLE_DUMP)
+void VPReverseVectorPointerRecipe::print(raw_ostream &O, const Twine &Indent,
+                                         VPSlotTracker &SlotTracker) const {
+  O << Indent;
+  printAsOperand(O, SlotTracker);
+  O << " = reverse-vector-pointer ";
```
Comment (fhahn) [inline llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:1849]: It looks like this is missing printing the inbounds flag. 

Same for VPVectorPointerRecipe, which should probably be fixed separately.

File: llvm/lib/Transforms/Vectorize/VPlan.h (hunk 5 above)
Comment (arcbbb) [reply llvm/lib/Transforms/Vectorize/VPlan.h]: Fixed. Thanks!

File: llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp (hunk 7 above)
Comment (arcbbb) [reply llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp]: Fixed. Thanks!

File: llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp (hunk 8 above)
Comment (arcbbb) [reply llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp]: Fixed. Thanks!

File: llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp (hunk 4 above)
Comment (arcbbb) [reply llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp]: Sure, I move them to getGEPIndexTy().

File: llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp [hunk 10]
```diff
@@ -1813,38 +1813,63 @@ void VPWidenGEPRecipe::print(raw_ostream &O, const Twine &Indent,
 }
 #endif
 
-void VPVectorPointerRecipe ::execute(VPTransformState &State) {
-  auto &Builder = State.Builder;
-  State.setDebugLocFrom(getDebugLoc());
-  unsigned CurrentPart = getUnrollPart(*this);
+static Type *getGEPIndexTy(bool IsScalable, bool IsReverse,
```
Comment (fhahn) [inline llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:1816]: Could you also add `VPReverseVectorPointerSC` to mayHaveSideEffects & co?

**Review by fhahn — APPROVED** [review]
LGTM, thanks.

But should probably fold the `mayHaveSideEffects ` change from https://github.com/llvm/llvm-project/pull/113410 into this PR
